kernel/src/table_changes/log_replay.rs

```diff
@@ -415,12 +424,14 @@ impl<'a> RowVisitor for FileActionSelectionVisitor<'a> {
                 continue;
             }
             if getters[0].get_str(i, "add.path")?.is_some() {
-                self.selection_vector[i] = !self.has_cdc_action;
-            } else if let Some(path) = getters[1].get_str(i, "remove.path")? {
+                let data_change: bool = getters[1].get(i, "add.dataChange")?;
+                self.selection_vector[i] = data_change && !self.has_cdc_action;
```

Comment:
Also: According to the Delta spec, CDC action doesn't have any stats for data skipping (just partition values). So maybe it's not helpful to apply a data skipping predicate in `has_cdc_action` case? Or maybe we leave the predicate in place, so it automatically picks up any future support for skipping over partition columns?

Reply:
I can do the simple optimization of checking `has_cdc_action` in the FileActionSelectionVisitor.  I think splitting the visitor could be done in a followup.

> CDC action doesn't have any stats for data skipping (just partition values)

I think it makes sense for the data skipping to remain if we can take advantage of partition columns as long as evaluating the predicate isn't too expensive.

Reply:
Implemented a version of this:
> we should check has_cdc_action first, to avoid invoking the getter unnecessarily?

I reordered the fields since it looked nicer.